### PR TITLE
Add Black compatible configurations in documentation (#1366 & #1205)

### LINKED ISSUE #1205: Document "black-compatible" configs for common linters/fixers
This is a documentation feature-request.

Commonly when bootstrapping a new project, my personal preferred config is (among other things): flake8, isort, black
The trouble is, every time I do this, I end up copy-pasting around the same config data so that flake8 and isort behave in a "black-compatible" way.

I also see issues when I occasionally run `pylint` on things because I don't personally use it often enough to have black-compatible config on-hand.

If `black` documented, in one place, config for commonly-used tools, I think that would be great. You can dig this info out of the ["The Black code style"](https://black.readthedocs.io/en/stable/the_black_code_style.html) page for `flake8`, `isort`, but
1. The information is kind of buried in there
2. It's not clear where to suggest additional config for other common tools

A while back, I reached out to [flake8 to ask about adding "black-compatible" mode as config there](https://gitlab.com/pycqa/flake8/issues/542), which is sort of pushing the responsibility for compatibility in the other direction. That feature request was rejected -- and I think the grounds for turning it down were reasonable.

But the thing I wanted there is still basically the same: I'm a user of several popular linters and fixers. I don't want, as a user, to be solely responsible for keeping these various tools compatible with one another. I want the projects to own this, and to be a consumer of that consolidated wisdom.

Fancy things -- like an option to generate config -- would be lovely, but that's not even necessary. Just having a single place to find docs which write down something like this would be great
<details><summary>setup.cfg</summary>

```
[isort]
multi_line_output = 3
include_trailing_comma = true
force_grid_wrap = 0
combine_as_imports = true
line_length = 88

[flake8]
max-line-length = 88
extend-ignore = W503,W504

[pylint]
# some stuff goes here, ignoring rules which break on black'd code

[common-tool-X-which-I'm-forgetting]
# more config
```
</details>

<details><summary>pyproject.toml</summary>

```
# compatible pyproject.toml content...
```
</details>

I'm happy to write the content myself, to the best of my ability, perhaps as a new docs page. But it doesn't seem worth doing unless it's actually wanted here.

### LINKED ISSUE #1366: Bring pylint to its senses with a .pylintrc
Black is better than pylint. It solves problems instead of just pointing them out. But pylint whines whenever black makes better choices. Can we get a sample .pylintrc file that will make pylint accept the wisdom of black's choices?

### PR DESCRIPTION
**Pull request by ichard26**
Resolves #1366 and resolves #1205.

**Please be warned, this is my first pull request and my first attempt at writing decent documentation.**
So feel free to mention if I've done anything wrong or if my documentation is lacking in quality.

~~Note that this PR isn't ready for review since some work needs to be finished:~~
- [x] I need to add references in README.  I haven't done it yet since I'm too scared of touching the README ... It's the front page of this project.
- [x] I should add explanations for the configurations since copying and pasting code you don't understand is always a horrible idea.

Also, while I did test all of the configurations to verify that they work with their respective tools, I may have missed a few options. I went for the least amount of configuration possible.

Once again, any comments/suggestions on my PR, my commit messages, or my actual documentation are welcome! Also so are comments/suggestions on how I should link from the README.

P.S: Thanks for this tool, it's really useful!

### CODE REVIEW

**Comment by ichard26** [comment]
b9146508bbcadd8bbf6a2fcbaeb2699fd58cea62 Whoops, looks like I missed changing my testing value. `max-line-length = 79` is now fixed to `max-line-length = 88`.

File: docs/compatible_configs.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+isort should be configured to match _Black_'s import formatting style by these options:
+
+- `multi_line_output = 3`
+- `include_trailing_comma = true`
+- `force_grid_wrap = 0`
+- `combine_as_imports = true`
+
+isort should be configured to wrap imports when they surpass _Black_'s default limit of
+88 characters via `line_length = 88`.
```
Comment (ichard26) [inline docs/compatible_configs.md:28]: I am not proud of this. It does a poor job of explaining the configuration. The problem is that I haven't used isort before and I don't know these options well enough to explain them well. Even reading the isort documentation didn't really do much for me.

Comment (ichard26) [reply docs/compatible_configs.md]: Well, after realizing my isort configuration was not ideal, I spent some more time reading the isort documentation and finally the options made sense. **Issue resolved.**

File: README.md [hunk 2]
```diff
@@ -127,6 +127,16 @@ _Black_ is a well-behaved Unix-style command-line tool:
 - it only outputs messages to users on standard error;
 - exits with code 0 unless an internal error occurred (or `--check` was used).
 
+### Using _Black_ with other tools
+
+While _Black_ enforces formatting that conforms to PEP 8, other tools may raise warnings
+about _Black_'s changes or will overwrite _Black_'s changes... I am looking at you
+[isort](https://pypi.org/p/isort). Since _Black_ is barely configurable, these tools
```
Comment (cooperlees) [inline README.md:134]: I don't think we should call out isort directly. It was never designed to work _with_ black. Can we remove this please or reword to be a little more subtle.
"A good example here is isort"

File: docs/compatible_configs.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
+# _Black_ compatible configurations
+
+Most of the changes that _Black_ make are harmless, but a few do conflict against other
```
Comment (JelleZijlstra) [inline docs/compatible_configs.md:3]: All of Black's changes are harmless (or at least, they should be)

File: docs/compatible_configs.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+```
+
+</details>
+
+## flake8
+
+[flake8](https://pypi.org/p/flake8/) is a code linter. It warns you of syntax errors,
+possible bugs, stylistic errors, etc. For the most part, flake8 follows
+[PEP 8](https://www.python.org/dev/peps/pep-0008/) when warning about stylistic errors.
+Except, there are a few deviations that cause incompatiblities with _Black_.
```
Comment (JelleZijlstra) [inline docs/compatible_configs.md:126]:
```suggestion
There are a few deviations that cause incompatibilities with _Black_.
```

File: docs/compatible_configs.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+When _Black_ is folding very long expressions, the closing brackets will
+[be dedented](https://github.com/psf/black#how-black-wraps-lines).
+
+```py3
+ImportantClass.important_method(
+    exc, limit, lookup_lines, capture_locals, callback
+)
+```
+
+Although, while this style is PEP 8 compliant, Pylint will raise
```
Comment (JelleZijlstra) [inline docs/compatible_configs.md:211]:
```suggestion
Although this style is PEP 8 compliant, Pylint will raise
```

File: docs/compatible_configs.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+multi_line_output = 3
+include_trailing_comma = True
+force_grid_wrap = 0
+use_parentheses = True
+line_length = 88
+```
+
+</details>
+
+## flake8
```
Comment (hugovk) [inline docs/compatible_configs.md:121]:
```suggestion
## Flake8
```

File: docs/compatible_configs.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+[flake8](https://pypi.org/p/flake8/) is a code linter. It warns you of syntax errors,
```
Comment (hugovk) [inline docs/compatible_configs.md:123]:
```suggestion
[Flake8](https://pypi.org/p/flake8/) is a code linter. It warns you of syntax errors,
```

File: docs/compatible_configs.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+possible bugs, stylistic errors, etc. For the most part, flake8 follows
```
Comment (hugovk) [inline docs/compatible_configs.md:124]:
```suggestion
possible bugs, stylistic errors, etc. For the most part, Flake8 follows
```

File: docs/compatible_configs.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+```
+
+### Why those options above?
+
+When breaking a line, _Black_ will break it before a binary operator. This is compliant
+with PEP 8, but this behaviour will cause flake8 to raise
+`W503 line break before binary operator` warnings.
+
+In some cases, as determined by PEP 8, _Black_ will enforce an equal amount of
+whitespace around slice operators. Due to this, flake8 will raise
```
Comment (hugovk) [inline docs/compatible_configs.md:142]:
```suggestion
whitespace around slice operators. Due to this, Flake8 will raise
```

File: docs/compatible_configs.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+`E203 whitespace before ':'` warnings.
+
+Since both of these warnings are not PEP 8 compliant, flake8 should be configured to
```
Comment (hugovk) [inline docs/compatible_configs.md:145]:
```suggestion
Since both of these warnings are not PEP 8 compliant, Flake8 should be configured to
```

File: docs/compatible_configs.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+ignore these warnings via `extend-ignore = E203, W503`.
+
+Also, as like with isort, flake8 should be configured to allow lines up to the length
```
Comment (hugovk) [inline docs/compatible_configs.md:148]:
```suggestion
Also, as like with isort, Flake8 should be configured to allow lines up to the length
```

File: docs/compatible_configs.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+[flake8]
+max-line-length = 88
+extend-ignore = E203, W503
+```
+
+</details>
+
+## Pylint
+
+[Pylint](https://pypi.org/p/pylint/) is also a code linter like flake8. It has the same
```
Comment (hugovk) [inline docs/compatible_configs.md:188]:
```suggestion
[Pylint](https://pypi.org/p/pylint/) is also a code linter like Flake8. It has the same
```

File: README.md (hunk 2 above)
Comment (ichard26) [reply README.md:134]: Yeah, I agree. In hindsight, isort and _Black_ were never designed to be compatible with each other and it's unfair to call out isort for that. I'll try to reword it to more gentle or just remove if I fail at doing that.

File: docs/compatible_configs.md (hunk 3 above)
Comment (ichard26) [reply docs/compatible_configs.md]: Well noted, I think I've been stalking the issue section a bit too much.

**Comment by cooperlees** [comment]
@ichard26 - I made the changes and merged! Thanks. We're having a black maintainers merge-a-thon today to try get black to become a python package - as per #1376

**Comment by ichard26** [comment]
Oh well, I was going to add a commit with your very well noted suggestions but looks like you beat me to it. Thank you for putting in time for PR  regardless! It was a good first experience for contributing to open source. 

> We're having a black maintainers merge-a-thon

Oh yeah I can tell, my poor little email is getting spammed... good sign though!
